core/new-gui/src/app/workspace/service/execute-workflow/execute-workflow.service.ts

```diff
@@ -0,0 +1,199 @@
+      requestURL,
+      JSON.stringify(body),
+      { headers: { 'Content-Type': 'application/json' } })
+      .subscribe(
+        // backend will either respond an execution result or an error will occur
+        // handle both cases
+        response => this.handleExecuteResult(response),
+        errorResponse => this.handleExecuteError(errorResponse)
+      );
+  }
+
+  /**
+   * Gets the observable for execution started event
+   * Contains a string that says:
+   *  - execution process has begun
+   */
+  public getExecuteStartedStream(): Observable<string> {
+    return this.executeStartedStream.asObservable();
+  }
+
+  /**
+   * Gets the observable for execution ended event
+   * If execution succeeded, it contains an object with type
+   *  `SuccessExecutionResult`:
+   *    -  resultID: the result ID of this execution
+   *    -  Code: the result code of 0
+   *    -  result: the actual result data to be displayed
+   *
+   * If execution succeeded, it contains an object with type
+   *  `ErrorExecutionResult`:
+   *    -  Code: the result code 1
+   *    -  message: error message
+   */
+  public getExecuteEndedStream(): Observable<ExecutionResult> {
+    return this.executeEndedStream.asObservable();
+  }
+
+  /**
+   * Handles valid execution result from the backend.
+   * Sends the execution result to the execution end event stream.
+   *
+   * @param response
+   */
+  private handleExecuteResult(response: SuccessExecutionResult): void {
+    console.log('handling success result ');
+    console.log(response);
+    this.executeEndedStream.next(response);
+  }
+
+  /**
+   * Handler function for invalid execution.
+   *
+   * Send the error messages generated from
+   *  backend (if workflow is invalid or server error)
+   *  or frontend (if there's no network connection)
+   *  to the execution end event stream.
+   *
+   * @param errorResponse
+   */
+  private handleExecuteError(errorResponse: HttpErrorResponse): void {
+    console.log('handling error result ');
+    console.log(errorResponse);
+
+    // error shown to the user in different error scenarios
+    const displayedErrorMessage = ExecuteWorkflowService.processErrorResponse(errorResponse);
+    this.executeEndedStream.next(displayedErrorMessage);
+  }
+
+  /**
+   * Transform a workflowGraph object to the HTTP request body according to the backend API.
+   *
+   * All the operators in the workflowGraph will be transformed to LogicalOperator objects,
+   *  where each operator has an operatorID and operatorType along with
+   *  the properties of the operator.
+   *
+   *
+   * All the links in the workflowGraph will be tranformed to LogicalLink objects,
+   *  where each link will store its source id as its origin and target id as its destination.
+   *
+   * @param workflowGraph
+   */
+  public static getLogicalPlanRequest(workflowGraph: WorkflowGraphReadonly): LogicalPlan {
+
+    const operators: LogicalOperator[] = workflowGraph
```

Comment:
For data members, as per my discussion with Zuozhi, type is not mandatory. @zuozhiw is it required here? And also check the **const links** data member below.

Reply:
In short it's not required here, but it's good to write the type explicitly. It's a more complicated to explain:

Here's a simpler example to illustrate this issue. Here we define an interface `A` has to members, `a` and `b`. 
```
interface A {
  a: number,
  b: number
}
```
We have a list `originalList`, we want to construct a `listOfA` by applying some mapper function:
```
const listOfA = originalList.map(item => {a: item.a, b: item.b});
```
We let TS infer that the type of listOfA could be `A[]`, however, consider what if we write the mapper function incorrectly:
```
const listOfA = originalList.map(item => {a: item.a});
```
Here we forgot the item b, although we still name the variable listOfA, we know that it no longer has type we wanted `A[]`, so in this case, it's better to write with type declaration, so that we can make sure the variable indeed has the type we wanted
```
// error
const listOfA: A[] = originalList.map(item => {a: item.a});
```